main.go

```diff
@@ -1,732 +1,717 @@
-		}
+			if storeMode == REMOVE {
+				if cacheModel.DebugMode {
+					log.Printf("  Exclude: %s", path)
+				}
+				return nil
+			}
 
-		fingerprintSourceAbsPth, err := pathutil.AbsPath(theFingerprintSourcePath)
-		if err != nil {
-			return []byte{}, fingerprintMeta, fmt.Errorf("Failed to get Absolute path of item (%s): %s", theFingerprintSourcePath, err)
-		}
+			if archiveFiles {
+				if err := cacheModel.TarWriter.WriteHeader(header); err != nil {
+					return err
+				}
+			}
 
-		fileInfo, isExist, err := pathutil.PathCheckAndInfos(fingerprintSourceAbsPth)
-		if err != nil {
-			return []byte{}, fingerprintMeta, fmt.Errorf("Failed to check the specified path: %s", err)
-		}
-		if !isExist {
-			return []byte{}, fingerprintMeta, errors.New("Specified path does not exist")
-		}
+			if info.IsDir() {
+				return nil
+			}
 
-		if fileInfo.IsDir() {
-			err := filepath.Walk(fingerprintSourceAbsPth, func(aPath string, aFileInfo os.FileInfo, walkErr error) error {
-				if walkErr != nil {
-					log.Printf(" (!) Error checking file (%s): %s", aPath, walkErr)
-				}
-				if aFileInfo.IsDir() {
-					// directory - skipping
-					return nil
-				}
-				if !aFileInfo.Mode().IsRegular() {
-					if gIsDebugMode {
-						log.Printf(" (i) File (%s) is not a regular file (it's a symlink, a device, or something similar) - skipping", aPath)
+			if header.Typeflag == tar.TypeReg {
+				switch storeMode {
+				case STORE:
+					if cacheModel.FileChangeIndicator == MD5 {
+						fileMD5, err := getFileContentMD5(path)
+						if err != nil {
+							return err
+						}
+						cacheModel.FilePathMap[path] = fileMD5
+					} else if cacheModel.FileChangeIndicator == MODTIME {
+						cacheModel.FilePathMap[path] = fmt.Sprintf("%d", info.ModTime().Unix())
 					}
-					return nil
-				}
-
-				if isShouldIgnorePathFromFingerprint(aPath, absIgnorePaths) {
-					if gIsDebugMode {
-						log.Printf(" [IGNORE] path from fingerprint: %s", aPath)
+				case SKIP:
+					if cacheModel.DebugMode {
+						log.Printf("  Ignore changes: %s", path)
 					}
-					fingerprintMeta[aPath] = FingerprintMetaModel{FingerprintSource: "-"}
-					isFingerprintGeneratedForPathItem = true
-					return nil
+					cacheModel.FilePathMap[path] = "-"
+				case INDICATOR:
+					cacheModel.FilePathMap[path] = indicatorFileMD5
 				}
 
-				fileFingerprintSource, err := fingerprintSourceStringOfFile(aPath, aFileInfo, fingerprintMethodID)
-				if err != nil {
-					return fmt.Errorf("Failed to generate fingerprint source for file (%s), error: %s", aPath, err)
-				}
+				if archiveFiles {
+					file, err := os.Open(path)
+					if err != nil {
+						return err
+					}
 
-				if gIsDebugMode {
-					log.Printf(" * fileFingerprintSource (%s): %#v", aPath, fileFingerprintSource)
-				}
-				fingerprintMeta[aPath] = FingerprintMetaModel{FingerprintSource: fileFingerprintSource}
+					_, err = io.CopyN(cacheModel.TarWriter, file, info.Size())
```

Comment:
if io.CopyN fails the file will not be closed, use defer, after `file, err := os.Open(path)`